Closes #1031: Remove non-regex substring search

pierce314159 opened the pull request:
Closes #1031:
Decreases size and complexity of substring search code by escaping non-regex patterns instead of doing specialized methods. Early benchmarking suggests this may be faster as well

NOTE: this would require users to compile chapel with regex support in order to use the substring search methods. I think that is reasonable especially if the resulting methods are faster

glitch reviewed (approved):
This looks good.

My only real question is whether or not we should add a check in the `Makefile` for forcing a check on whether Chapel was built with Regex support.  We could probably do something similar to how the ZMQ & HDF5 checks are done (i.e. compile a chapel program which tries to use re2).  Alternatively `chpl --help-settings | grep re2 -a1` yields
```
      --re2 <re2-version>             Specify RE2 library
          currently: ''
```
Which could be a lighter weight way to do it with some further output processing.

pierce314159 commented on arkouda/strings.py:
```diff
@@ -678,21 +678,20 @@ def contains(self, substr: Union[bytes, str_scalars], regex: bool = False) -> pd
-                                             json.dumps([substr]))
+        if not regex:
+            substr = re.escape(substr)
+        if re.search(substr, ''):
```
ah yeah the reason I do this is because I'm checking for patterns that match the empty string which is a subtle distinction from being equal to the empty string

so this also catches cases like `substr='$'` or `substr='|'` which still match the empty string and would cause the server to hang for users running anything less than the latest version of chapel

pierce314159 commented:
> This looks good.
> 
> My only real question is whether or not we should add a check in the `Makefile` for forcing a check on whether Chapel was built with Regex support. We could probably do something similar to how the ZMQ & HDF5 checks are done (i.e. compile a chapel program which tries to use re2). Alternatively `chpl --help-settings | grep re2 -A1` yields
> 
> ```
>       --re2 <re2-version>             Specify RE2 library
>           currently: ''
> ```
> 
> Which could be a lighter weight way to do it with some further output processing.
> 
> EDIT: one-liner but it's not pretty
> 
> ```
> if [[ -z `chpl --help-settings |grep re2 -A1|grep currently|cut -d ":" -f2|tr -d \'|tr -d [:space:]` ]]; then echo "no re2"; fi
> ```

yeah this sounds like a good idea to me! If we're gonna require users have regex support, having it be an explicit check in the makefile seems like the right way to go

pierce314159 commented:
I updated the chapel install instruction in the `README` to include `export CHPL_RE2=bundled`

I also added a check to the makefile to test that chapel is built with `RE2` support based on the hdf and zmq checks. I verified the check fails when the `CHPL_RE2` flag is not set

`checkRE2.chpl` checks the version to ensure we use the right regex module. This does result in a deprecation warning when running make. I think this should be fine since it does use the correct module for the version, but I'm curious what others think

bradcray commented on dep/checkRE2.chpl:
```diff
@@ -0,0 +1,14 @@
+    use Regex;
+    var myRegex = compile("a+");
+  }
+  writeln("Chapel correctly made with RE2 support\n");
```
Noted on gitter but capturing here:  Unfortunately, this doesn't work as intended because the Chapel compiler today parses all `use` and `import` statements regardless of the scope they appear in.  So this will attempt to parse both `Regexp` and `Regex` which works for Chapel 1.25.x because both modules exist there (`Regexp` as a deprecation placeholder that will be removed soon), but will not work for Chapel 1.24.x because `Regex` didn't exist there yet.

Building Arkouda itself does work because of `src/compat/lt-125/ArkoudaRegexCompat.chpl` which handles this in a different way.
